Assets/Scripts/EOSTransferTask.cs

```diff
@@ -0,0 +1,305 @@
+                PlayerDataStorageService.Log($"{nameof(EOSTransferTask)} ({nameof(OnFileTransferProgressUpdated)}): Player data storage: transfer progress {data.BytesTransferred} / {data.TotalFileSizeBytes}");
+
+                OnProgressUpdated?.Invoke(Progress);
+            }
+        }
+
+        /// <summary>
+        /// Method that is called when a file transfer download has completed.
+        /// </summary>
+        /// <param name="data">EOS SDK data about the file transfer.</param>
+        internal void OnFileReceived(ref ReadFileCallbackInfo data)
+        {
+            if (data.ResultCode != Result.Success)
+            {
+                ResultCode = data.ResultCode;
+                InnerTaskCompletionSource.SetException(new Exception($"{nameof(EOSTransferTask)} ({nameof(OnFileReceived)}): Could not receive data. Result code {data.ResultCode}"));
+                return;
+            }
+
+            ResultCode = Result.Success;
+            InnerTaskCompletionSource.SetResult(Data);
+        }
+
+        /// <summary>
+        /// Internal processor for writing chunks of data during file upload.
+        /// </summary>
+        /// <param name="data">Metadata about the file upload.</param>
+        /// <param name="outDataBuffer">Output buffer for this next chunk of data.</param>
+        /// <returns>A command to either continue or finish writing data.</returns>
+        internal WriteResult OnFileDataSend(ref WriteFileDataCallbackInfo data, out ArraySegment<byte> outDataBuffer)
+        {
+            outDataBuffer = new ArraySegment<byte>();
+
+            if (IsDone())
+            {
+                return WriteResult.CompleteRequest;
+            }
+
+            uint bytesToWrite = Math.Min(MAX_CHUNK_SIZE, TotalSize - currentIndex);
+
+            if (bytesToWrite > 0)
+            {
+                outDataBuffer = new ArraySegment<byte>(Data, (int)currentIndex, (int)bytesToWrite);
+            }
+
+            currentIndex += bytesToWrite;
+
+            if (IsDone())
+            {
+                return WriteResult.CompleteRequest;
+            }
+            else
+            {
+                return WriteResult.ContinueWriting;
+            }
+        }
+
+        /// <summary>
+        /// Method for handling when a file upload has completed.
+        /// </summary>
+        /// <param name="data">EOS SDK data about the file transfer.</param>
+        internal void OnFileSent(ref WriteFileCallbackInfo data)
+        {
+            if (data.ResultCode != Result.Success)
+            {
+                ResultCode = data.ResultCode;
+                InnerTaskCompletionSource.SetException(new Exception($"{nameof(EOSTransferTask)} ({nameof(OnFileSent)}): Could not send data. Result code {data.ResultCode}"));
+                return;
+            }
+
+            ResultCode = Result.Success;
+            InnerTaskCompletionSource.SetResult(Data);
+        }
```

Comment:
I think these function names need to be improved. It's not clear to me what the difference is between `OnFileSent` and `OnFileDataSent`, since (without looking at their respective implementations) they _mean_ the same thing. We should strive to have our method names be clearer, and not require a developer to look at the implementation (or even read the comments) to understand the distinction between the two.